File: ChatGPT(예진)/Python (수정 완료)/BlackJack_Game.py
import random

# Initialize card deck
suits = ["Spades", "Diamonds", "Hearts", "Clubs"]
ranks = ["2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"]
values = [2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 11]

# Define card class
class Card:
    def __init__(self, suit, rank):
        self.suit = suit
        self.rank = rank

# Create card deck
class Deck:
    def __init__(self):
        self.deck = [Card(suit, rank) for suit in suits for rank in ranks]
        self.current_card = 0

    def shuffle_deck(self):
        random.shuffle(self.deck)

    def deal_card(self):
        card = self.deck[self.current_card]
        self.current_card += 1
        return card

# Define hand class
class Hand:
    def __init__(self):
        self.cards = []
        self.num_cards = 0
        self.value = 0
        self.aces = 0

    def add_card_to_hand(self, card):
        self.cards.append(card)
        self.num_cards += 1
        card_value_index = ranks.index(card.rank)
        self.value += values[card_value_index]
        if card.rank == "A":
            self.aces += 1

    def adjust_for_ace(self):
        while self.value > 21 and self.aces:
            self.value -= 10
            self.aces -= 1

# Player's turn
def player_turn(deck, player_hand):
    while True:
        print("\nPlayer's cards:")
        for card in player_hand.cards:
            print(f"{card.rank} of {card.suit}")
        print(f"Player's card total: {player_hand.value}")

        if player_hand.value == 21:
            print("Blackjack! Player wins!")
            break
        elif player_hand.value > 21:
            print("Card total exceeds 21. Player busts!")
            break
        else:
            action = input("Would you like to hit or stay? (hit/stay): ")
            if action == "hit":
                player_hand.add_card_to_hand(deck.deal_card())
                player_hand.adjust_for_ace()
            elif action == "stay":
                break

# Dealer's turn
def dealer_turn(deck, dealer_hand):
    while dealer_hand.value < 17:
        dealer_hand.add_card_to_hand(deck.deal_card())
        dealer_hand.adjust_for_ace()

# Play the game
def blackjack_game():
    while True:
        # Initialize game
        deck = Deck()
        deck.shuffle_deck()

        player_hand = Hand()
        dealer_hand = Hand()

        player_hand.add_card_to_hand(deck.deal_card())
        player_hand.add_card_to_hand(deck.deal_card())
        dealer_hand.add_card_to_hand(deck.deal_card())
        dealer_hand.add_card_to_hand(deck.deal_card())

        # Player's turn
        player_turn(deck, player_hand)

        # Dealer's turn
        if player_hand.value <= 21:
            dealer_turn(deck, dealer_hand)

        # Print results
        print("\nDealer's cards:")
        for card in dealer_hand.cards:
            print(f"{card.rank} of {card.suit}")
        print(f"Dealer's card total: {dealer_hand.value}")

        if dealer_hand.value > 21:
            print("Dealer busts! Player wins!")
        elif dealer_hand.value > player_hand.value:
            print("Dealer wins!")
        elif dealer_hand.value < player_hand.value:
            print("Player wins!")
        else:
            print("It's a tie!")

        # Check if the player wants to play again
        play_again = input("\nDo you want to play again? (yes/no): ")
        if play_again.lower() != "yes":
            break

if __name__ == "__main__":
    blackjack_game()

File: ChatGPT(예진)/C (수정 완료)/BlackJack_Game.c
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

// Initialize card deck
char* suits[] = { "Spades", "Diamonds", "Hearts", "Clubs" };
char* ranks[] = { "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A" };
int values[] = { 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10, 11 };

// Define card structure
typedef struct {
    char* suit;
    char* rank;
} Card;

// Create card deck
typedef struct {
    Card deck[52];
    int current_card;
} Deck;

// Define hand structure
typedef struct {
    Card cards[11]; // Assume a hand can hold a maximum of 11 cards
    int num_cards;
    int value;
    int aces;
} Hand;

// Initialize card deck
void initialize_deck(Deck* deck) {
    int i, j, k = 0;
    for (i = 0; i < 4; i++) {
        for (j = 0; j < 13; j++) {
            deck->deck[k].suit = suits[i];
            deck->deck[k].rank = ranks[j];
            k++;
        }
    }
    deck->current_card = 0;
}

// Shuffle the deck
void shuffle_deck(Deck* deck) {
    int i, j;
    Card temp;
    for (i = 0; i < 52; i++) {
        j = rand() % 52;
        temp = deck->deck[i];
        deck->deck[i] = deck->deck[j];
        deck->deck[j] = temp;
    }
}

// Deal a card
Card deal_card(Deck* deck) {
    return deck->deck[deck->current_card++];
}

// Initialize hand
void initialize_hand(Hand* hand) {
    hand->num_cards = 0;
    hand->value = 0;
    hand->aces = 0;
}

// Add a card to hand
void add_card_to_hand(Hand* hand, Card card) {
    hand->cards[hand->num_cards++] = card;
    int i;
    for (i = 0; i < 13; i++) {
        if (strcmp(card.rank, ranks[i]) == 0) {
            hand->value += values[i];
            if (strcmp(card.rank, "A") == 0) // Handle ace
                hand->aces++;
            break;
        }
    }
}

// Adjust for ace
void adjust_for_ace(Hand* hand) {
    while (hand->value > 21 && hand->aces) {
        hand->value -= 10;
        hand->aces--;
    }
}

// Player's turn
void player_turn(Deck* deck, Hand* player_hand) {
    char action[10];
    while (1) {
        printf("\nPlayer's cards:\n");
        for (int i = 0; i < player_hand->num_cards; i++) {
            printf("%s of %s\n", player_hand->cards[i].rank, player_hand->cards[i].suit);
        }
        printf("Player's card total: %d\n", player_hand->value);

        if (player_hand->value == 21) {
            printf("Blackjack! Player wins!\n");
            break;
        }
        else if (player_hand->value > 21) {
            printf("Card total exceeds 21. Dealer wins!\n");
            break;
        }
        else {
            printf("Would you like to hit or stay? (hit/stay): ");
            scanf("%s", action);
            if (strcmp(action, "hit") == 0) {
                add_card_to_hand(player_hand, deal_card(deck));
                adjust_for_ace(player_hand);
            }
            else if (strcmp(action, "stay") == 0) {
                break;
            }
        }
    }
}

// Dealer's turn
void dealer_turn(Deck* deck, Hand* dealer_hand) {
    while (dealer_hand->value < 17) {
        add_card_to_hand(dealer_hand, deal_card(deck));
        adjust_for_ace(dealer_hand);
    }
}

// Play the game
void blackjack_game() {
    srand(time(NULL));

    while (1) {
        // Initialize game
        Deck deck;
        initialize_deck(&deck);
        shuffle_deck(&deck);

        Hand player_hand, dealer_hand;
        initialize_hand(&player_hand); // Initialize player's hand
        initialize_hand(&dealer_hand);

        add_card_to_hand(&player_hand, deal_card(&deck));
        add_card_to_hand(&player_hand, deal_card(&deck));
        add_card_to_hand(&dealer_hand, deal_card(&deck));
        add_card_to_hand(&dealer_hand, deal_card(&deck));

        // Player's turn
        player_turn(&deck, &player_hand);

        // Dealer's turn
        if (player_hand.value <= 21) {
            dealer_turn(&deck, &dealer_hand);
        }

        // Print results
        printf("\nDealer's cards:\n");
        for (int i = 0; i < dealer_hand.num_cards; i++) {
            printf("%s of %s\n", dealer_hand.cards[i].rank, dealer_hand.cards[i].suit);
        }
        printf("Dealer's card total: %d\n", dealer_hand.value);

        if (dealer_hand.value > 21) {
            printf("Dealer busts! Player wins!\n");
        }
        else if (dealer_hand.value > player_hand.value) {
            printf("Dealer wins!\n");
        }
        else if (dealer_hand.value < player_hand.value) {
            printf("Player wins!\n");
        }
        else {
            printf("It's a tie!\n");
        }

        // Check if the player wants to play again
        char play_again[4];
        printf("\nDo you want to play again? (yes/no): ");
        scanf("%s", play_again);
        if (strcmp(play_again, "no") == 0) {
            break;
        }
    }
}

int main() {
    blackjack_game();
    return 0;
}